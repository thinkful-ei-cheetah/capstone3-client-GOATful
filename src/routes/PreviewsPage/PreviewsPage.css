.previews-page {
  display: flex;
}

.previews-display-section {
  padding: 2em;
  padding-top:0;
  padding-bottom:0;
  margin: auto;
}

@media screen and (max-width: 630px) {
  .previews-page {
    flex-flow: row wrap;
    justify-content: center;
  }
}