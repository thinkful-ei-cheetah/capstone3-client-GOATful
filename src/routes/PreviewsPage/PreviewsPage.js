import React, { Component } from 'react'
import {  Redirect } from 'react-router-dom'
import VideoStorage from '../../services/video-storage'
import PreviewControls from '../../components/PreviewControls/PreviewControls'
import PreviewsApi from '../../services/previews-api'
import './PreviewsPage.css'
import { withAppContext } from '../../contexts/AppContext';
import CreatorPreview from '../../components/CreatorPreview/CreatorPreview'
import FAB from '../../components/FAB/FAB'
import YoutubeApiService from '../../services/youtube-api'
import mockYoutubeData from '../../Utils/mock-youtube-date'
import VideoApi from '../../services/video-api'

class Previews extends Component {
  static defaultProps = {
    match: {
      params: {
        video_id: 0
      }
    }
  }
  constructor(props) {
    super(props)
    this.state = {
      video: {},
      vidPreviews: [],
      activePrev:{},
      selectedPrev: null,
      userPreview: {},
      youtubeSearchResults: [],
      isLoading: true,
      isAdd:false,
    }
    this.vidId = Number(this.props.match.params.video_id);
  }

  findSelect = (prevObj) => {
    let select = false
    let active = this.findActive(prevObj)
    if (active !== false) {
      select = active
    } else {
      select = prevObj.previews[0]
    }
    return select
  }

  findActive = (prevObj) => {
    let active = false;
    let temp = prevObj.previews.find(preview => {
      return preview.is_active === true
    })
    if (temp !== undefined) {
      active = temp
    } else {
      active = false
    }
    return active
  }

  async componentDidMount() {
    try {
      const prevObj = await PreviewsApi.getPreviews(this.vidId)
      VideoStorage.saveKey('laconic_current_video', prevObj.video);
      let activePreview = this.findActive(prevObj);
      let selected = this.findSelect(prevObj);
      let previews = [...prevObj.previews]
      if(activePreview){
        let activeIndex = previews.findIndex((prev) => {
          return prev.is_active;
        })
        let activePreview = previews[activeIndex]
        previews[activeIndex] = previews[0]
        previews[0] = activePreview
      }

      let youtubeSearchResults;
      try {
        youtubeSearchResults = await VideoApi.getYoutubeSearchResults(this.vidId)
        if (!youtubeSearchResults.length) {
          youtubeSearchResults = await YoutubeApiService.search(prevObj.video.tags)
          await VideoApi.postYoutubeSearchResults(this.vidId, youtubeSearchResults)
        }
      } catch(err) {
        youtubeSearchResults = mockYoutubeData
      }
      this.setState({
        video: prevObj.video,
        vidPreviews: previews,
        activePrev: activePreview,
        selectedPrev: selected,
        youtubeSearchResults: youtubeSearchResults,
        isLoading: false
      })
    } catch(err) {
      this.setState({isLoading: false}, this.props.appContext.setAppError(`An error occurred while trying to load your video previews, the server responded with: ${err.message}`))
    }
  }

  renderPreviews = () => {
    const { selectedPrev, video } = this.state
    video.title = selectedPrev.title
    video.thumbnail_url = selectedPrev.thumbnail_url
    video.description = selectedPrev.description

    return <CreatorPreview userPreview={{ ...video }} setLoading={() => {}}/>
  }

  previewClick = (e) => {
    e.preventDefault();
    let selected = this.state.vidPreviews.find(preview => {
      return preview.id === parseInt(e.target.id)
    })
    
    this.setState({
      selectedPrev: selected
    })
  }

  editClick = (e) => {
    VideoStorage.saveKey('laconic_current_preview', { ...this.state.selectedPrev })
  }

  delClick = async () => {
    this.setState({isLoading: true})
    try {
      await PreviewsApi.deletePreview(parseInt(this.vidId), this.state.selectedPrev.id)
      const updatedPreviews = await PreviewsApi.getPreviews(this.vidId)
      if (updatedPreviews.previews.length === 0) {
        this.props.history.push('/creator')
      } else {
        const selected = this.findSelect(updatedPreviews)
        this.setState({
          selectedPrev: selected,
          vidPreviews: updatedPreviews.previews,
          isLoading: false
        })
      }
    } catch (error) {
      this.setState({isLoading: false}, this.props.appContext.setAppError(`An error occurred while trying to delete your video preview - the server responded with: ${error.message}`))
    }
  }

  handleFab = () => {
    this.setState({
      isAdd: true
    })
  }

  handleTouchStart = () => {
    document.body.style.overflow = 'auto';
  }

  handleSetActive = async () => {
    let preview = {...this.state.selectedPrev};
    preview.is_active = true;
    preview.changeActive = true;
    await PreviewsApi.patchPreview(this.vidId, preview)

    this.setState({
      activePrev: {...preview}
    })
  }

  render() {
    return (
      <section className="previews-page page">
        <FAB onClick={this.handleFab} />
        <PreviewControls
          prevList={this.state.vidPreviews}
          selected={this.state.selectedPrev}
          activePrev={this.state.activePrev}
          previewClick={this.previewClick}
          editClick={this.editClick}
          delClick={this.delClick}
          setActive={this.handleSetActive}
        />
        <div className="previews-display-section" onTouchStart={this.handleTouchStart}>
          {(this.state.selectedPrev === null) ? false : this.renderPreviews()}
        </div>
        {(this.state.isAdd)? <Redirect to='/creator' />: false}
      </section>
    );
  }
}

export default withAppContext(Previews)