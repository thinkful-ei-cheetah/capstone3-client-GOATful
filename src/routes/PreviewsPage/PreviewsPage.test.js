import React from 'react';
import ReactDOM from 'react-dom';
import PreviewsPage from './PreviewsPage';
import renderer from 'react-test-renderer';


describe('PreviewsPage Component', () => {
  it('renders without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(< PreviewsPage />, div);
    ReactDOM.unmountComponentAtNode(div);
  });

  it('renders the UI as expected', () => {
    const tree = renderer
      .create(<PreviewsPage/>)
      .toJSON();
    expect(tree).toMatchSnapshot();  
  });
})
