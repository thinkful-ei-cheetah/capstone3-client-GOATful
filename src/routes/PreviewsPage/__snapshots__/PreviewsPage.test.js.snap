// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PreviewsPage Component renders the UI as expected 1`] = `
<section
  className="previews-page page"
>
  <button
    className="fab"
    onClick={[Function]}
  >
    <svg
      aria-hidden="true"
      className="svg-inline--fa fa-plus fa-w-14 fab-icon"
      data-icon="plus"
      data-prefix="fas"
      focusable="false"
      onClick={[Function]}
      role="img"
      style={Object {}}
      viewBox="0 0 448 512"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M416 208H272V64c0-17.67-14.33-32-32-32h-32c-17.67 0-32 14.33-32 32v144H32c-17.67 0-32 14.33-32 32v32c0 17.67 14.33 32 32 32h144v144c0 17.67 14.33 32 32 32h32c17.67 0 32-14.33 32-32V304h144c17.67 0 32-14.33 32-32v-32c0-17.67-14.33-32-32-32z"
        fill="currentColor"
        style={Object {}}
      />
    </svg>
  </button>
  <div
    className="preview-controls-container"
  >
    <h2
      className="preview-title"
    >
      Previews
    </h2>
    <div
      className="preview-tiles"
      id="hScroll"
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
      onWheel={[Function]}
    />
  </div>
  <div
    className="previews-display-section"
    onTouchStart={[Function]}
  />
</section>
`;
